pkg/controllers/provisioningv2/rke2/machinestatus/controller.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/rancher/rancher/pkg/wrangler"
+	"github.com/rancher/wrangler/pkg/condition"
+	corecontrollers "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/rancher/wrangler/pkg/relatedresource"
+	corev1 "k8s.io/api/core/v1"
+	apierror "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	capi "sigs.k8s.io/cluster-api/api/v1alpha4"
+)
+
+const (
+	Provisioned         = condition.Cond("Provisioned")
+	InfrastructureReady = condition.Cond(capi.InfrastructureReadyCondition)
+	BootstrapReady      = condition.Cond(capi.BootstrapReadyCondition)
+)
+
+type handler struct {
+	secrets              corecontrollers.SecretCache
+	machines             capicontrollers.MachineController
+	bootstrapCache       rkecontroller.RKEBootstrapCache
+	bootstrapController  rkecontroller.RKEBootstrapController
+	capiClusterCache     capicontrollers.ClusterCache
+	provClusterCache     provisioningcontrollers.ClusterCache
+	mgmtClusterCache     mgmtcontrollers.ClusterCache
+	rkeControlPlaneCache rkecontroller.RKEControlPlaneCache
+	dynamic              *dynamic.Controller
+}
+
+func Register(ctx context.Context, clients *wrangler.Context) {
+	h := handler{
+		secrets:              clients.Core.Secret().Cache(),
+		machines:             clients.CAPI.Machine(),
+		bootstrapCache:       clients.RKE.RKEBootstrap().Cache(),
+		bootstrapController:  clients.RKE.RKEBootstrap(),
+		capiClusterCache:     clients.CAPI.Cluster().Cache(),
+		mgmtClusterCache:     clients.Mgmt.Cluster().Cache(),
+		provClusterCache:     clients.Provisioning.Cluster().Cache(),
+		rkeControlPlaneCache: clients.RKE.RKEControlPlane().Cache(),
+		dynamic:              clients.Dynamic,
+	}
+	clients.CAPI.Machine().OnChange(ctx, "machine-status", h.OnChange)
+
+	relatedresource.Watch(ctx, "machine-trigger-from-secret", func(namespace, name string, obj runtime.Object) ([]relatedresource.Key, error) {
+		if secret, ok := obj.(*corev1.Secret); ok {
+			if secret.Type == planner.SecretTypeMachinePlan {
+				return []relatedresource.Key{{
+					Namespace: secret.Namespace,
+					Name:      secret.Labels[planner.MachineNameLabel],
+				}}, nil
+			}
+		}
+		return nil, nil
+	}, clients.CAPI.Machine(), clients.Core.Secret())
+
+	h.dynamic.OnChange(ctx, "machine-trigger", func(gvk schema.GroupVersionKind) bool {
+		return gvk.Group == "rke-machine.cattle.io"
+	}, func(obj runtime.Object) (runtime.Object, error) {
+		m, err := meta.Accessor(obj)
+		if err != nil {
+			return nil, err
+		}
+		for _, owner := range m.GetOwnerReferences() {
+			if owner.Kind == "Machine" {
+				h.machines.Enqueue(m.GetNamespace(), owner.Name)
+			}
+		}
+		return obj, nil
+	})
+}
+
+func (h *handler) setJoinURLFromOutput(machine *capi.Machine, nodePlan *plan.Node) error {
+	if nodePlan == nil || !planner.IsEtcdOnlyInitNode(machine) || machine.Annotations[planner.JoinURLAnnotation] != "" {
```

Comment:
`!planner.IsEtcdOnlyInitNode(machine)`, so if the node is not an etcd-only init node, we just return nil? What if it's a controlplane+etcd init node?

What happens if the join URL changes for this node, i.e. if the node IP changes? We won't update the JoinURLAnnotation?

Reply:
First, I'll not that this code hasn't changed in the PR. It was just moved to a different file.

This function is only meant to handle the case where the init-node is an etcd only node. In every other instance, the join URL is set in the planner.

Reply:
Thanks for the clarification! And sorry, I didn't realize it was just moved haha